Mathlib/RingTheory/MvPolynomial/Symmetric.lean

```diff
@@ -309,4 +309,244 @@ theorem psum_isSymmetric (n : ℕ) : IsSymmetric (psum σ R n) := rename_psum _
+
+instance exponentFinite (n : ℕ) : Finite (Exponent σ n) := by
+  let f : Exponent σ n → (σ → Fin (n+1)) := fun s => fun i => ⟨s.exp i, Nat.lt_succ_of_le (Exponent.single_le_sum σ s i)⟩
+  have hf : Function.Injective f := by
+    intros s t h
+    rw [Exponent.ext_iff]
+    ext i
+    have := congr_arg (fun s => s i) h
+    simp only [Function.Injective.eq_iff] at this
+    exact Fin.mk_eq_mk.mp this
+  exact Finite.of_injective f hf
+  done
+
+instance exponentFintype (n : ℕ) : Fintype (Exponent σ n) := Fintype.ofFinite (Exponent σ n)
+
+def exponentLiftEquiv (e : σ ≃ τ) : Exponent σ n ≃ Exponent τ n := {
+          toFun := fun s => ⟨fun i => s.exp (e.invFun i), by simp_rw [← s.sum_eq]; exact Equiv.sum_comp e.symm s.exp⟩,
+          invFun := fun s => ⟨fun i => s.exp (e.toFun i), by simp_rw [← s.sum_eq]; exact Equiv.sum_comp e s.exp⟩,
+          left_inv := fun s => by simp,
+          right_inv := fun s => by simp
+          }
+
+lemma exponentLiftEquiv_toFun (e : σ ≃ τ) (s : Exponent σ n) (i : τ) : ((exponentLiftEquiv _ e).toFun s).exp i = s.exp (e.invFun i) := by
+  simp only [Equiv.toFun_as_coe, Equiv.invFun_as_coe]
+  rfl
+  done
+
+lemma Exponent.exponentCard_zero : Fintype.card (Exponent σ 0) = 1 := by
+  apply Fintype.card_eq_one_iff.mpr
+  use Exponent.zero σ
+  intro x
+  ext i
+  simp only [Exponent.sum_eq, Exponent.zero_of_sum_eq_zero]
+  done
+
+lemma Exponent.one_sum_singleton (s : Exponent σ 1) : ∃! (i : σ) , Exponent.one σ i = s:= by
+  have h : ∑ i, s.exp i = 1 := s.sum_eq
+  have (i : σ): exp s i = 0 ∨ exp s i = 1 := by
+    have := Exponent.single_le_sum σ s i
+    apply Nat.le_one_iff_eq_zero_or_eq_one.mp
+    exact this
+  by_cases exists_i : ∃ (i : σ), s.exp i = 1
+  · obtain ⟨i, hi⟩ := exists_i
+    use i
+    constructor
+    simp only [Exponent]
+    ext j
+    by_cases hji : j = i
+    · rw [hji, Exponent.one]
+      simp only [ite_true, hi]
+    · have : exp s j = 0 := by
+        obtain (h0 | h1) := (this j)
+        · assumption
+        · exfalso
+          have : 2 <= ∑ i, s.exp i := by
+            let f : Finset σ := {i, j}
+            have : f ⊆ univ := subset_univ f
+            have that : ∑ k in f, s.exp k = 2 := by
+              simp only [Finset.sum_insert]
+              have : ∀ k ∈ f, s.exp k = 1 := by
+                intro k hk
+                have : k = i ∨ k = j := by
+                  simp only [Finset.mem_insert] at hk
+                  obtain (ki | kj) := hk
+                  · left
+                    exact ki
+                  · right
+                    exact mem_singleton.mp kj
+                obtain (ki | kj) := this
+                · rw [ki]
+                  exact hi
+                · rw [kj]
+                  exact h1
+              have f2 : f.card = 2 := by
+                simp only [Finset.card_insert_of_not_mem, Finset.card_singleton, hji]
+                rw [Finset.card_eq_two]
+                use i, j
+                constructor
+                symm
```

Comment:
Same as below: you (presumably) have multiple goals here; please use focusing dots.